src/proc_diskstats.c

```diff
@@ -69,21 +79,45 @@ struct disk *get_disk(unsigned long major, unsigned long minor) {
 		last->next = d;
 	}
 
+	// ------------------------------------------------------------------------
+	// find the type of the device
+
 	// find if it is a partition
-	// by reading /sys/dev/block/MAJOR:MINOR/partition
+	// by checking if /sys/dev/block/MAJOR:MINOR/partition is readable.
 	char buffer[FILENAME_MAX + 1];
-	snprintf(buffer, FILENAME_MAX, path_find_block_device_partition, major, minor);
+	snprintf(buffer, FILENAME_MAX, path_find_block_device, major, minor, "partition");
+	if(access(buffer, R_OK) == 0)
+		d->type = DISK_TYPE_PARTITION;
+
+	// find if it is a container
+	// by checking if /sys/dev/block/MAJOR:MINOR/slaves has entries
+	if(d->type == DISK_TYPE_UNKNOWN) {
+		int is_container = 0;
+		snprintf(buffer, FILENAME_MAX, path_find_block_device, major, minor, "slaves/");
+		DIR *dirp = opendir(buffer);	
+		if (dirp != NULL) {
+		struct dirent *dp;
+			for(;;) {
+				dp = readdir(dirp);
+
+				if(dp == NULL) {
+					break;
+				}
+				if(strcmp(dp->d_name, ".") == 0 || strcmp(dp->d_name, "..") == 0) {
+					continue;
+				}
 
-	int fd = open(buffer, O_RDONLY, 0666);
-	if(likely(fd != -1)) {
-		// we opened it
-		int bytes = read(fd, buffer, FILENAME_MAX);
-		close(fd);
+				d->type = DISK_TYPE_CONTAINER;
```

Comment:
> Do I realy need to add two braces around while((dp = readdir(dirp)))? What does that do?

Without them the compiler will complain: Did the developer forgot an equal sign, or he/she really means to make dp = readdir?